[Docs] Add Twig Extensions examples (#8397)
pimcore/pimcore: doc/Development_Documentation/02_MVC/02_Template/02_Template_Extensions/README.md

# Twig Extensions

## Introduction

Twig Extensions are functions, filters and more that offer special functionality to increase usability of view scripts. 

Following an overview of some [Twig Extensions](https://twig.symfony.com/doc/2.x/#reference):   
- `render` 
- `render_esi` 
- `controller` 
- `asset` 
- `csrf_token`  
- `path` 
- `url` 
- `absolute_url` 
- `translator` 
- `trans`
- `url` 
- `relative_path` 

  
In addition to the standard Twig extensions, Pimcore adds some additional powerful extensions. 
  
## Pimcore Twig Extensions

All Twig extension functions are described below in detail, the following tables give just a short overview of all available extensions.

| Extension                                | Description                                                     |
|------------------------------------------|-----------------------------------------------------------------|

> | `pimcore_action()`        | (Deprecated) Call an arbitrary action and renders the respective template  |

Reviewer: ```suggestion
| `pimcore_action()`        | (Deprecated) Calls an arbitrary action and renders the respective template  |
```

| `pimcore_cache()`         | Simple in-template caching functionality                                         |

> | `pimcore_device()`        | Helps implementing adaptive designs.                                             |

Reviewer: ```suggestion
| `pimcore_device()`        | Helps implementing adaptive designs                                              |
```

| `pimcoreglossary()`       | Extension to control the glossary engine                                         |
| `pimcore_placeholder()`   | Adding and embedding custom placeholders, e.g. for special header tags, etc.     |
| `pimcore_head_link()`     | Embeding / managing referenced stylesheets (alternative to `assets()`)           |
| `pimcore_head_meta()`     | Managing your \<meta\> elements in your HTML document                            |

> | `pimcore_head_script()`   | Managing your <scripts> elements                                                 |

Reviewer: ```suggestion
| `pimcore_head_script()`   | Managing your \<scripts\> elements                                               |
```

| `pimcore_head_style()`    | Managing inline styles (pendant to `headLink()` for inline styles)               |
| `pimcore_head_title()`    | Create and store the HTML document's `<title>` for later retrieval and output    |

> | `pimcore_inc()`           | Use this function to directly include a Pimcore document.                        |

Reviewer: ```suggestion
| `pimcore_inc()`           | Use this function to directly include a Pimcore document                         |
```

| `pimcore_inline_script`   | Managing inline scripts (pendant to `headScript()` for inline scripts)           |
| `pimcore_build_nav()`, `pimcore_render_nav()`, `pimcore_nav_renderer()`   | Embed and build navigations based on the document structure                             |

> | `pimcore_url()`           | An alternative to `url()` and `path()` with the building behavior of Pimcore 4   |

Reviewer: `with the building behavior of Pimcore 4` is this relevant?

Author: kicked it

> | `pimcore_website_config()`| Fetch website settings or specific setting(first param: key) for the current site |

Reviewer: ```suggestion
| `pimcore_website_config()`| Fetch website settings or specific setting (first param: key) for the current site |
```

> | `pimcore_image_thumbnail()` | Returns a path to a given thumbnail on image                                   |

Reviewer: ```suggestion
| `pimcore_image_thumbnail()` | Returns a path to a given thumbnail or image                                   |
```

Author: this is called on image

> | `pimcore_image_thumbnail_html()` | Returns html for displaying the thumbnail image.                          |

Reviewer: ```suggestion
| `pimcore_image_thumbnail_html()` | Returns html for displaying the thumbnail image                           |
```

> | `pimcore_supported_locales()` | Use this function to get list of support locales.                            |

Reviewer: ```suggestion
| `pimcore_supported_locales()` | Use this function to get a list of supported locales                         |
```

> | `pimcore_image_thumbnail()` | Returns a path to a given thumbnail or an thumbnail configuration.             |

Reviewer: ```suggestion
| `pimcore_image_thumbnail()` | Returns a path to a given thumbnail or thumbnail configuration                 |
```

> | `pimcore_image_thumbnail()` | Returns a path to a given thumbnail or an thumbnail configuration.             |

Reviewer: ```suggestion
| `pimcore_image_thumbnail()` | Returns a path to a given thumbnail or thumbnail configuration                 |
```

Pimcore also adds some Twig tests for evaluating boolean conditions e.g.
```twig
{# using 'instaceof' checks if object is instanceof provided classname #}
{% if( product is instanceof('AppBundle\\Model\\Product\\Car')) %}
    ...
{% endif %}

{# using 'pimcore_data_object' checks if object is instanceof \Pimcore\Model\DataObject\Concrete #}

> {% if( product is pimcore_data_object) %}

Reviewer: ```suggestion
{% if (product is pimcore_data_object) %}
```

 ...
{% endif %}
```

> The following tables give an overview of all available tests:

Reviewer: ```suggestion
The following table gives an overview of all available tests:
```

|--------------|--------------|--------------|
| `default`    | string       | Default if no device can be detected |

##### Example
```twig
{% if pimcore_device().isPhone() %}
    This is my phone content
{% elseif pimcore_device().isTablet() %}
    This text is shown on a tablet
{% elseif pimcore_device().isDesktop() %}
    This is for default desktop Browser
{% endif %}
```
   
For details also see [Adaptive Design](../../../19_Development_Tools_and_Details/21_Adaptive_Design_Helper.md).

### `pimcoreglossary`

The `pimcoreglossary` block replaces glossary terms. See [Glossary](../../../18_Tools_and_Features/21_Glossary.md) for details.

```twig
{% pimcoreglossary %}
My content
{% endpimcoreglossary %}
``` 

### `pimcore_inc` 
Use `pimcore_inc()` to include documents (eg. snippets) within views. 
This is especially useful for footers, headers, navigations, sidebars, teasers, ...

`pimcore_inc(document, params, cacheEnabled)`

| Name           | Type         | Description  |
|----------------|--------------|--------------|
| `document`     | PageSnippet &#124; int &#124; string  | Document to include, can be either an ID, a path or even the Document object itself |

> | `params`       |  array       | Is optional and should be an array with key value pairs like in `$this->action()` from ZF. |

Reviewer: `...from ZF.` there is no Zend anymore ;)

Author: kicked it

thanks for your review 🙏

| `enabledCache` |  bool        | Is true by default, set it to false to disable the cache. Hashing is done across source and parameters to ensure a consistent result. |
 

>  ##### Example

Reviewer: ```suggestion
##### Example
```

Used to interact with navigations. See [Navigation](../../../03_Documents/03_Navigation.md) for details. Simplified example:

```twig
{% set navigation = pimcore_build_nav({
    active: document,
    root: navRootDocument
}) %}
{{ pimcore_render_nav(navigation) }}

{# you can also fetch the renderer instance and call custom render methods #}
{% set renderer = pimcore_nav_renderer('menu') %}
{{ renderer.render(navigation) }}
```

### `pimcore_url` 
An alternative to `url()` and `path()` which used Url. 

```twig
{{ pimcore_url(params, name, reset, encode, relative) }}
```

All parameters are optional here:

| Name         | Type   | Description  |
|--------------|--------|--------------|
| `params`     | array  | Route params. If object is passed in the params then link generator will be used to generate Url  |
| `name`       | string | Route name |
| `reset`      | bool   | Is false by default, set it to false to avoid merging parameters from request  |
| `encode`     | bool   | Is true by default, set it to false to disable encoding |
| `relative`   | bool   | Is false by default, set it to true to generate a relative path based on the current request path |

>  ##### Example

Reviewer: ```suggestion
##### Example
```
